man/koopa.1.ronn
# koopa(1) - shell bootloader for bioinformatics

## SYNOPSIS

`koopa` ...

## OPTIONS

### Unix standard (general)

  * `--help`, `-h`:
    Show this help message and exit.

  * `--version`, `-V`:
    Version information.

### System configuration

  * `check`:
    Run system dependency version checks.

  * `develop`:
    Switch koopa installation to developer mode.

  * `info`:
    Show system information.

  * `list`:
    List programs accessible in `PATH`.

  * `update`:
    Update koopa installation to latest version.
    Run with `--system` flag to also update the system configuration.

### Character string return

  * `app-prefix`:
    Application installation prefix.
    Refers to applications that don't need to be build using `make`, and can be
    installed on an external disk, for space-constrained virtual machines.
    Defaults to "/n/opt/" (with external disk) or "/usr/local/opt" (without
    external disk) on most machines.

  * `config-prefix`:
    User-specific local configuration prefix.
    Defined inside `XDG_CONFIG_HOME`, following recommended XDG configuration.

  * `header`:
    Shell script header file path.
    Can be passed to `source` inside external shell scripts.
    Useful for experimental scripts that are not currently saved inside koopa.

  * `host-id`:
    General platform identifier string.
    Intended for identification of cloud platforms, such as
    `aws` (Amazon Web Services), `azure` (Microsoft Azure),
    or `gcp` (Google Cloud Platform).

  * `local-app-prefix`:
    Local application prefix.
    Defaults to `XDG_DATA_HOME` on most machines.

  * `make-prefix`:
    Prefix for `make` application build scripts.
    Defaults to "/usr/local" on most machines.

  * `os-string`:
    Operating system identifier string.
    Returns OS name and version, separated by a hyphen.
    Example: "ubuntu-18" for Ubuntu 18 LTS or "rhel-8" for RHEL 8.

  * `prefix`:
    Return the koopa installation path prefix.
    Also stored as `KOOPA_PREFIX` environment variable.

  * `r-home`:
    R installation prefix (`R_HOME`).

  * `variable`:
    Internal koopa configuration variable.
    Recommended for use in Python and R scripts only.

## COPYRIGHT

This software is provided under the GNU General Public License v3.0.
See `LICENSE` file for details.


NEWS
## koopa 0.3.2 (2019-04-18)

- Improved build scripts, and hardened some intended for RedHat Linux.
- Updated genome build information.
- Improved bcbio scripts and custom genome builds.
- New exported scripts: `ip-address.sh`, `vim-sort.sh`.
- Added script to build Python from source for RedHat.



## koopa 0.3.1 (2019-03-19)

- Updated to Ensembl release to 95.
- Updated GENCODE to release human 29 / mouse M20.
- Updated FlyBase to 2019_01.
- `git-pull-all.sh`: Added `git fetch` call.
- Improved build configuration scripts: emacs, git, gsl, openssl, tmux.
- Added bcbio system configuration files (bcbio_system.yaml).
- Improved default run template for bcbio RNA-seq.



## koopa 0.3.0 (2019-02-20)

This release overhauls the previous activation method. Refer to the README
for details on how to reconfigure your shell profile.

- Changed activation method to use `activate.sh` script at top of repo.
- Improved support for ksh and zsh shells. Added Travis CI checks.
- Added native support for zsh pure prompt, including async script. These files
  are saved in `zsh/fpath/`.
- Reorganized bcbio configuration scripts and templates.
- Completely overhauled system activation method. Additional shell
  configuration, including prompt string customization is now disabled by
  default. Refer to the installation instructions for details on how to enable
  this new optional `extra` mode.
- Reorganized installer scripts exported in `bin/` including conda scripts,
  `reset-permissions.sh`, `sudo-yum-update.sh`, etc.



## koopa 0.2.8 (2019-02-12)

- Migrated darwin (macOS)-specific scripts from bash repo to koopa.
- Migrated Harvard O2 and Harvard Odyssey HPC scripts here from bash repo.
- Added `rsync-azure-files.sh` script for working with Azure Shares (Samba).
- Added `autopad-samples.sh` script.
- Added `bash-strict-mode.sh` script.
- Added `md5sum.sh` script.
- Reorganized other scripts accessible via `bin/`.
- Reorganized activation and system scripts, in preparation for 0.3 release.



## koopa 0.2.7 (2019-02-04)

- Improved shellchecks on workflows.
- Split out some activation steps into separate scripts.



## koopa 0.2.6 (2019-01-23)

- Added back ".sh" extension for all exported scripts.
- Updated Travis CI configuration to use shellcheck.
- Reorganized and improved workflow scripts.



## koopa 0.2.5 (2019-01-07)

- Updated installation instructions to recommend cloning to `~/.koopa`.
- Conda now will only activate when `CONDA_DEFAULT_ENV` is set in environment.
  This improves handling in situations for user accounts where we don't want
  to activate conda in a shared environemnt (e.g. bioinfo account on Azure).



## koopa 0.1.1 (2018-09-05)

- Now exporting `KOOPA_VERSION` and `KOOPA_DATE` global variables.
- Updated installation instructions.
- Removed `/setup/` config files.
- Added support for automatic SSH key import, using `$SSH_KEY` variable.
- Removed "g*" aliases for GitHub commands (e.g. ga, gc, gd). Too confusing.
- Simplified the separator bar for koopa message, avoiding unicode.
- Renamed `bcl2fastq_indrop` script to `bcl2fastq_indrops`.
- Added some utility scripts from macOS bash repo:
      - delete_fs_cruft
      - git_pull_all
      - gzip_dir
      - install_tex
      - move_files_up_1_level
      - rename_from_csv
      - reset_permissions



## koopa 0.1.0 (2018-08-24)

- Initial stable release.
- Repo was previously named "seqcloud".


etc/completion/koopa.sh
#!/usr/bin/env bash

# """
# Bash/Zsh TAB completion.
# Updated 2020-11-18.
#
# Keep all of these commands in a single file.
# Sourcing multiple scripts doesn't work reliably.
#
# Multi-level bash completion:
# - https://stackoverflow.com/questions/17879322/
# - https://stackoverflow.com/questions/5302650/
# """

# FIXME NEED TO PROVIDE AUTOCOMPLETE SUPPORT FOR INSTALL, UPDATE, UNINSTALL

_koopa_complete() {
    local args cur prev
    COMPREPLY=()
    cur=${COMP_WORDS[COMP_CWORD]}
    prev=${COMP_WORDS[COMP_CWORD-1]}
    if [ "$COMP_CWORD" -eq 1 ]
    then
        args=(
            '--help'
            '--version'
            'app'
            'check-system'
            'get-version'
            'header'
            'info'
            'install'
            'list'
            'prefix'
            'test'
            'uninstall'
            'update'
        )
        if _koopa_is_linux
        then
            args+=(
                'delete-cache'
            )
        fi
        COMPREPLY=("$(compgen -W "${args[*]}" -- "$cur")")
    elif [ "$COMP_CWORD" -eq 2 ]
    then
        case "$prev" in
            app)
                args=(
                    'clean'
                    'list'
                    'link'
                    'unlink'
                )
                COMPREPLY=("$(compgen -W "${args[*]}" -- "$cur")")
                ;;
            header)
                args=(
                    'bash'
                    'posix'
                    'r'
                    'zsh'
                )
                COMPREPLY=("$(compgen -W "${args[*]}" -- "$cur")")
                ;;
            install)
                args=(
                    ## > 'mike'
                    'dotfiles'
                    'py-koopa'
                    'r-koopa'
                )
                COMPREPLY=("$(compgen -W "${args[*]}" -- "$cur")")
                ;;
            # FIXME REWORK THIS.
            # FIXME DROPPED TO SYSTEM.
            get-version)
                args=(
                    'emacs'
                    'vim'
                )
                COMPREPLY=("$(compgen -W "${args[*]}" -- "$cur")")
                ;;
            system)
                args=(
                    'log'
                    'pull'
                )
                COMPREPLY=("$(compgen -W "${args[*]}" -- "$cur")")
                ;;
            update)
                args=(
                    '--fast'
                )
                COMPREPLY=("$(compgen -W "${args[*]}" -- "$cur")")
                ;;
            *)
                ;;
        esac
    fi
    return 0
}
complete -F _koopa_complete koopa
